composer/models/deeplabv3/model.py

```diff
@@ -133,8 +141,20 @@ def deeplabv3(num_classes: int,
             else:
                 model.apply(initializer_fn)
 
-    if sync_bn:
-        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
+    if sync_bn and world_size > 1:
+        local_world_size = dist.get_local_world_size()
+
+        # List of ranks for each node, assumes that each node has the same number of ranks
+        num_nodes = world_size // local_world_size
+        process_group = None
+        if num_nodes > 1:
```

Comment:
instead of creating all the process groups for each node, and then selecting based on the node_rank, can this be simplified to:
```
node_rank = dist.get_node_rank()
ranks = list(range(node_rank * local_world_size, (node_rank + 1) * local_world_size))
process_group = torch.dist.new_group(ranks)
...
```
```

Reply:
Yeah, that should work haha. I think I tried to follow the pytorch example too closely, thanks!

Reply:
Hmmm, I'm getting a weird NCCL error with the change. Maybe torch.distributed needs to know that all of the processes are assigned to certain groups even in each process? Not really sure.

```
Traceback (most recent call last):
  File "/composer/examples/run_composer_trainer.py", line 78, in <module>
    _main()
  File "/composer/examples/run_composer_trainer.py", line 74, in _main
    trainer.fit()
  File "/usr/local/lib/python3.9/dist-packages/composer/trainer/trainer.py", line 1354, in fit
    self._train_loop()
  File "/usr/local/lib/python3.9/dist-packages/composer/trainer/trainer.py", line 1521, in _train_loop
    total_loss = self._train_batch(use_grad_scaling)
  File "/usr/local/lib/python3.9/dist-packages/composer/trainer/trainer.py", line 1664, in _train_batch
    total_loss = self._train_microbatches(microbatches)
  File "/usr/local/lib/python3.9/dist-packages/composer/trainer/trainer.py", line 1761, in _train_microbatches
    self._train_microbatch(use_grad_scaling, current_batch_size, total_loss, is_final_microbatch)
  File "/usr/local/lib/python3.9/dist-packages/composer/trainer/trainer.py", line 1798, in _train_microbatch
    self.state.outputs = self.state.model(self.state.batch)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/parallel/distributed.py", line 1008, in forward
    output = self._run_ddp_forward(*inputs, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/parallel/distributed.py", line 971, in _run_ddp_forward
    return module_to_run(*inputs, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/composer/models/tasks/classification.py", line 89, in forward
    outputs = self.module(inputs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/composer/models/deeplabv3/model.py", line 35, in forward
    features = self.backbone(x)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torchvision/models/_utils.py", line 69, in forward
    x = module(x)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/batchnorm.py", line 748, in forward
    return sync_batch_norm.apply(
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/_functions.py", line 56, in forward
    dist._all_gather_base(combined_flat, combined, process_group, async_op=False)
  File "/usr/local/lib/python3.9/dist-packages/torch/distributed/distributed_c10d.py", line 2136, in _all_gather_base
    work = group._allgather_base(output_tensor, input_tensor)
RuntimeError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1191, unhandled system error, NCCL version 2.10.3
ncclSystemError: System call (e.g. socket, malloc) or external library call failed or device error. It can be also caused by unexpected exit of a remote peer, you can check NCCL warnings for failure reason and see if there is connection closure by a peer.
```

Reply:
odd, ok. this is just a nit, so i'm OK merging in its current form.